src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4393,3 +4393,181 @@ void C2_MacroAssembler::vector_maskall_operation32(KRegister dst, Register src,
+
+  vpand(dst, dst, src, vec_enc);
+  vpsllq(dst, dst, 4, vec_enc);
+  vpand(xtmp2, xtmp2, src, vec_enc);
+  vpsrlq(xtmp2, xtmp2, 4, vec_enc);
+  vpor(xtmp1, dst, xtmp2, vec_enc);
+
+  movl(rtmp, 0x33333333);
+  movq(dst, rtmp);
+  vpbroadcastd(dst, dst, vec_enc);
+  vpxor(xtmp2, xtmp3, dst, vec_enc);
+
+  vpand(dst, dst, xtmp1, vec_enc);
+  vpsllq(dst, dst, 2, vec_enc);
+  vpand(xtmp2, xtmp2, xtmp1, vec_enc);
+  vpsrlq(xtmp2, xtmp2, 2, vec_enc);
+  vpor(xtmp1, dst, xtmp2, vec_enc);
+
+  movl(rtmp, 0x55555555);
+  movq(dst, rtmp);
+  vpbroadcastd(dst, dst, vec_enc);
+  vpxor(xtmp2, xtmp3, dst, vec_enc);
+
+  vpand(dst, dst, xtmp1, vec_enc);
+  vpsllq(dst, dst, 1, vec_enc);
+  vpand(xtmp2, xtmp2, xtmp1, vec_enc);
+  vpsrlq(xtmp2, xtmp2, 1, vec_enc);
+  vpor(dst, dst, xtmp2, vec_enc);
+
+  if (bt != T_BYTE) {
+    vector_reverse_byte_avx(bt, dst, dst, xtmp1, vec_enc);
+  }
+}
+
+void C2_MacroAssembler::vector_reverse_bit_evex(BasicType bt, XMMRegister dst, XMMRegister src, XMMRegister xtmp1,
+                                                XMMRegister xtmp2, Register rtmp, int vec_enc) {
+  // Shift based bit reversal.
+  movl(rtmp, 0x0f0f0f0f);
+  evpbroadcastd(xtmp1, rtmp, vec_enc);
+  vpternlogd(xtmp2, 0x11, xtmp1, xtmp1, vec_enc);
+
+  vpandq(dst, xtmp1, src, vec_enc);
+  vpsllq(dst, dst, 4, vec_enc);
+  vpandq(xtmp2, xtmp2, src, vec_enc);
+  vpsrlq(xtmp2, xtmp2, 4, vec_enc);
+  vporq(xtmp1, dst, xtmp2, vec_enc);
+
+  movl(rtmp, 0x33333333);
+  evpbroadcastd(dst, rtmp, vec_enc);
+  vpternlogd(xtmp2, 0x11, dst, dst, vec_enc);
+
+  vpandq(dst, dst, xtmp1, vec_enc);
+  vpsllq(dst, dst, 2, vec_enc);
+  vpandq(xtmp2, xtmp2, xtmp1, vec_enc);
+  vpsrlq(xtmp2, xtmp2, 2, vec_enc);
+  vporq(xtmp1, dst, xtmp2, vec_enc);
+
+  movl(rtmp, 0x55555555);
+  evpbroadcastd(dst, rtmp, vec_enc);
+  vpternlogd(xtmp2, 0x11, dst, dst, vec_enc);
+
+  vpandq(dst, dst, xtmp1, vec_enc);
+  vpsllq(dst, dst, 1, vec_enc);
+  vpandq(xtmp2, xtmp2, xtmp1, vec_enc);
+  vpsrlq(xtmp2, xtmp2, 1, vec_enc);
+  vporq(dst, dst, xtmp2, vec_enc);
+
+  if (bt != T_BYTE) {
+    vector_reverse_byte_evex(bt, dst, dst, xtmp1, xtmp2, rtmp, vec_enc);
+  }
+}
+
+#ifdef _LP64
+void C2_MacroAssembler::vector_reverse_bit_gfni(BasicType bt, XMMRegister dst, XMMRegister src,
+                                                XMMRegister xtmp, Register rtmp, int vec_enc) {
+  // Galois field instruction based bit reversal.
+  assert(VM_Version::supports_gfni(), "");
+  mov64(rtmp, 0x8040201008040201L);
+  movq(xtmp, rtmp);
+  vpbroadcastq(xtmp, xtmp, vec_enc);
+  vgf2p8affineqb(dst, 0, src, xtmp, vec_enc);
+  if (bt != T_BYTE) {
+    if (VM_Version::supports_avx512bw()) {
+      vector_reverse_byte_evex(bt, dst, dst, xtmp, xnoreg, noreg, vec_enc);
+    } else {
+      assert(vec_enc < Assembler::AVX_512bit, "");
+      vector_reverse_byte_avx(bt, dst, dst, xtmp, vec_enc);
+    }
+  }
+}
+#endif
+
+void C2_MacroAssembler::vector_reverse_byte_avx(BasicType bt, XMMRegister dst, XMMRegister src,
```

Comment:
Since this is an in-lane shuffle, can we just use `vpshufb` for this?
Thanks.

Reply:
We are doing an element-wise byte reverse not a whole vector byte reverse I think we don't need any cross-lane shuffle here.
Thanks.

Reply:
Correct!, I realized it after commenting, corrected my statement.

Reply:
But a 3 cycle sequence still looks better than loading from memory followed by shuffle unless we are saving on code size as depicted in VBMI sequence which offers a direct vpermb instruction.

Reply:
Removed incorrect comments. your comments seems valid here.